---
title: "All Samples"
author: "Kennedi Todd"
date: "2/15/2021"
output:
pdf_document: default
html_document: default
---

Set working directory.
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/m214960/Documents/R_Workspace/Allen_Brain_Atlas/raw_data")
getwd()
```

Load necessary packages.
```{r}
library(BiocParallel)
library(data.table)
library(edgeR)
library(ggplot2)
library(ggrepel) # geom_text_repel()
library(gplots) # heatmap.2()
library(grid)
library(gridExtra)
library(knitr)
library(limma)
library(MatchIt)
library(rtracklayer)
library(variancePartition)
```

Read in files.
```{r}
counts <- read.delim("hippocampus_tpm_counts.txt", header = TRUE, sep = "\t")
gene.info <- read.delim("gene_info_for_counts_file.txt", header = TRUE, sep = "\t")
metadata <- read.csv("metadata.txt", header = TRUE)
```


```{r}
# Check that row names of counts file matches gene_id in gene.info
all.equal(as.vector(rownames(counts)),as.vector(as.character(gene.info$gene_id)))

# Returned true so rename rows with gene_symbol
row.names(counts) <- gene.info$gene_symbol

# Get rid of columns not useful
gene.info$score <- NULL
gene.info$phase <- NULL

# Check that column names of counts file matches X_donor_id in metadata
all.equal(as.vector(colnames(counts)), as.vector(metadata$X_donor_id))

# Get rid of columns not useful
metadata$X <- NULL
metadata$lib.size <- NULL
metadata$norm.factors <- NULL
```

Create DGE object with our counts, gene.info and metadata.
```{r}
dge <- DGEList(counts = counts, genes = gene.info, samples = metadata)
```

Only some of the ages are binned.  Let's bin the rest of the data.
```{r}
table(dge$samples$age)
```


```{r}
donor_ages <- dge$samples$age
bin_ages <- vector()

for (i in 1:length(donor_ages)){
  x <- donor_ages[i]
  if (x == "77" | 
      x == "78" | 
      x == "79") {
        bin_ages <- c(bin_ages, "75_79")
  }
  else if (x == "81" | 
      x == "82" | 
      x == "83" | 
      x == "84") {
        bin_ages <- c(bin_ages, "80_84")
  }
  else if (x == "85" | 
      x == "86" | 
      x == "87" | 
      x == "88" | 
      x == "89") {
        bin_ages <- c(bin_ages, "85_89")
  }
  else if (x == "85" | 
      x == "86" | 
      x == "87" | 
      x == "88" | 
      x == "89") {
        bin_ages <- c(bin_ages, "85_89")
  }
  else if (x == "90-94") {
        bin_ages <- c(bin_ages, "90_94")
  }
  else if (x == "95-99") {
        bin_ages <- c(bin_ages, "95_99")
  }
  else {
    bin_ages <- c(bin_ages, "100_")
  }
}
```


```{r}
as.data.frame(cbind(donor_ages,bin_ages)) #check
```

```{r}
dge$samples$age <- factor(bin_ages)
```

Reformat type of sample columns.
```{r}
sample_columns <- c(1, 7:24)
dge$samples[sample_columns] <- 
  lapply(dge$samples[sample_columns], factor)
```

-----------------------------TTR TPM HISTOGRAMS---------------------------------

Data is already in TPM (transcripts per million).
Create histogram of TPM TTR values.
```{r}
ttr.counts <- as.data.frame(dge$counts["TTR",])
colnames(ttr.counts) <- "ttr.counts"
log2.ttr.counts <- log2(ttr.counts + 0.01)
colnames(log2.ttr.counts) <- "log2.ttr.counts"
```

```{r}
h1 <- ggplot(ttr.counts, aes(x = ttr.counts)) + 
  geom_histogram(bins = 100, fill = "gray", color = "black") + 
  labs(title = "Allen Brain Atlas - Hippocampus", x=NULL, y=NULL) +
  xlab("TTR TPM") + ylab("# of Samples") + 
  geom_vline(xintercept = 20, col = "red") +
  theme_bw()

h2 <- ggplot(ttr.counts, aes(x = ttr.counts)) + 
  geom_histogram(bins = 100, fill = "gray", color = "black") + 
  labs(title = "Allen Brain Atlas - Hippocampus", x=NULL, y=NULL) +
  xlab("TTR TPM") + ylab("# of Samples") +
  xlim(0, 50) + geom_vline(xintercept = 20, col = "red") + 
  theme_bw() 

h3 <- ggplot(log2.ttr.counts, aes(x = log2.ttr.counts)) + 
  geom_histogram(bins = 100, fill = "gray", color = "black") + 
  labs(title = "Allen Brain Atlas - Hippocampus", x=NULL, y=NULL) +
  xlab("log2(TTR TPM)") + ylab("# of Samples") +
  geom_vline(xintercept = log2(20), col = "red") + theme_bw() +
  geom_vline(xintercept = log2(20) + 1, col = "blue") +
  geom_vline(xintercept = log2(20) - 1, col = "blue")
```


Arrange graphs in grid.
```{r}
plots1 <- list(h1,h2,h3)

layout1 <- rbind(c(1,2),c(3))

grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
```

------------------------------ASSIGN DIRTY/CLEAN/NEITHER------------------------

Lets looks at groups +/- 1 of the cutoff
```{r}
clean_threshold <- log2(20) - 1 # values <= this are clean
clean_threshold
cutoff <- log2(20)
cutoff
dirty_threshold <- log2(20) + 1 # values >= this are dirty
dirty_threshold
```

Assign clean or dirty
```{r}
clean_or_dirty <- vector()
vector.log2.ttr.counts <- as.vector(log2.ttr.counts$log2.ttr.counts)

for (i in 1:length(vector.log2.ttr.counts)){
  if (vector.log2.ttr.counts[i] <= clean_threshold){
    clean_or_dirty <- c(clean_or_dirty, "clean")
  }
  else if (vector.log2.ttr.counts[i] >= dirty_threshold){
    clean_or_dirty <- c(clean_or_dirty, "dirty")
  }
  else {
    clean_or_dirty <- c(clean_or_dirty, "neither")
  }
}
table(clean_or_dirty)
```


```{r}
dge$samples$group <- factor(clean_or_dirty)
head(dge$samples)
```

```{r}
h4 <- h3 + 
  annotate("rect", 
              xmin = -Inf,
              xmax = clean_threshold, 
              ymin = 0, 
              ymax=Inf, 
              alpha=0.2, 
              fill="deepskyblue") +
  annotate("rect", 
              xmin = dirty_threshold,
              xmax = Inf, 
              ymin = 0, 
              ymax=Inf, 
              alpha=0.2, 
              fill="chocolate4")
h4
```


Write table.
```{r}
write.csv(as.data.frame(dge$samples),"metadata_with_groups.txt")
```


-----------------------FILTER LOWLY EXPRESSED GENES-----------------------------

TPM to CPM
```{r}
cpm <- cpm(dge)
```

Filter
```{r}
# produces a logical matrix
thresh <- cpm > 2 # cpm greater than 2

# keep genes that have at least 3 TRUES in each row of thresh
keep <- rowSums(thresh) >= 3
summary(keep)

# redefine object, filter by logical
dge.filtered <- dge[keep, keep.lib.sizes = FALSE]

dim(dge)
dim(dge.filtered)
```
Went from 50,283 genes to 21,679 genes.

--------------------------TRIMMED MEAN OF M-VALUES------------------------------

Now, we want to normalize gene expression distributions. We do this by normalizing 
library size differences in each sample. The method Trimmed Mean of M-values (TMM) 
is used with the calcNormFactors in the edgeR package. The normalization factors 
calculated here are used as a scaling factor for the library size.
  
Trimmed mean of M values (TMM) normalization estimates sequencing
depth after excluding genes for which the ratio of counts between a pair
of experiments is too extreme or for which the average expression is too
extreme. The edgeR software implements a TMM normalization (Biostar Handbook 2020).

```{r}
# creates another large DGEList 
dge.filtered.norm <- calcNormFactors(dge.filtered, method = "TMM")

# view norm factors
dge.filtered.norm$samples$norm.factors
summary(dge.filtered.norm$samples$norm.factors)
```
Looking to be around ~1.

-------------------------------Variance Partition-------------------------------

See what options we have to choose from.
```{r}
colnames(dge.filtered.norm$samples)
```


Specify variables to consider
```{r}
register(SnowParam(4)) # work in parallel

# geneExpr: matrix of gene expression values
# info: information/metadata about each sample
geneExpr <- as.matrix(dge.filtered.norm$counts)
info <- as.data.frame(dge.filtered.norm$samples)

# Age is USUALLY continuous so model it as a fixed effect "age"
# However, this metadata categorizes it into bins
# Sex is categorical, so model them as random effects "(1|sex)"
# Note the syntax
form <- ~ (1|group) + 
  (1|age) + 
  (1|sex) + 
  (1|apo_e4_allele) +
  (1|cerad) +
  (1|dsm_iv_clinical_diagnosis) +
  (1|nincds_arda_diagnosis) +
  (1|act_demented) +
  (1|braak) +
  (1|nia_reagan) +
  (1|hemisphere)

varPart <- fitExtractVarPartModel(geneExpr, form, info)
vp <- sortCols(varPart)
```

Violin plots
```{r}
plotVarPart(vp)
```

```{r}
table(dge.filtered$samples$dsm_iv_clinical_diagnosis)
```


Percent bars
```{r}
plotPercentBars(vp[1:10,])
```

```{r}
# sort genes based on variance explained by group
varPart.df <- as.data.frame(varPart)
order.varPart.df <- varPart.df[order(varPart.df$group, decreasing = TRUE),]
order.varPart.df["group"]

```


-----------------------------DESIGN MATRIX--------------------------------------

```{r}
sex <- dge.filtered.norm$samples$sex
age <- dge.filtered.norm$samples$age
```


Create a design matrix and specify the samples we want to compare to each other.
```{r}
# interaction computes a factor which represents the interaction of the given factors
group <- interaction(dge.filtered$samples$group)

design <- model.matrix(~ 0 + group + sex + age)
colnames(design)[1:2] <- c("clean","dirty")
```

Run voom with quality weights. This combines observational-level weights with
  sample-specific quality weights in an experiment.
Normalizes expression intensities so that the log-ratios have similar distributions
  across a set of samples.
To quantile normalize, add normalize.method = "quantile".

Quality weights in RNA-seq data improves data quality when samples with high
  variance are present.  This variation is modeled by taking into account
  global intensity-dependent trends using voom.  
Sample-specific variability is taken into account with a log-linear model that 
  shares parameters between genes.
https://doi.org/10.1093/nar/gkv412


```{r}
v <- voomWithQualityWeights(dge.filtered.norm, design, plot = TRUE)
```

---------------------------------MDS PLOT---------------------------------------

```{r}
#create a png file# png(filename = "mds_1and2.png", width = 400, height = 400)
#default height and width are 480, default units are "px" (pixels)

group_colors <- c("deepskyblue","chocolate4","gray")[v$targets$group] # blue is clean, brown is dirty
point_shapes <- c("F","M")[v$targets$sex] # will say F or M for points
head(data.frame(v$targets$group, group_colors, v$targets$sex,point_shapes)) # Check everything matches

mds <- plotMDS(
  v, # our data object
  top = 10, # only looking at top 10 genes
  pch = point_shapes,
  cex = 1, # point size
  dim.plot = c(1,2), # specifying principal components to be plotted
  plot = TRUE, # if TRUE then plot is created on current graphics device
  col = group_colors # assigning our colors for each genotype
)

legend(
  "top",
  pch = 16,
  legend = unique(factor(v$targets$group)),
  col = group_colors,
  cex = 0.8
)

title("Raw Data: Top 10 Genes")
```



-------------------------HIERARCHIAL HEAT CLUSTERING----------------------------

ds[order(factor(Month, levels = c("Apr", "May", "Jan", "Feb", "Mar"))), .SD,ID]
```{r}
current_order <- dge.filtered.norm$samples$X_donor_id

# reorder levels
group_order <- dge.filtered.norm$samples
group_order <- group_order[order(factor(group, levels = c("dirty","neither","clean"))),]
sample_group_order <- group_order$X_donor_id
dge.group <- dge.filtered.norm[,sample_group_order]

# sample group colors
sample_group_color <- c("deepskyblue", "chocolate4", "gray")[dge.group$samples$group]
as.data.frame(dge.group$samples$group, sample_group_color)
```



```{r}
# We estimate the variance for each row in the logcounts matrix
logcounts <- cpm(dge.group, log = TRUE) # log2 values
var_genes <- apply(logcounts, 1, var)

# Get the gene names for the top 100 most variable genes
select_var <- names(sort(var_genes, decreasing=TRUE))[1:50]
select_var

# Subset logcounts matrix
highly_variable_lcpm <- logcounts[select_var,]
```


```{r}

my_palette <- colorRampPalette(c("red", "black", "green"))

# Plot the heatmap
map <- heatmap.2(highly_variable_lcpm, 
          main = "Top 50 most variable genes across samples",
          col = my_palette, # heatmap color scale
          ColSideColors = sample_group_color,
          trace = "none",
          dendrogram = "row",
          Colv = sample_group_order,
          colCol = sample_group_color,
          srtCol = 45,
          xlab = "Samples"
          )
```


---------------------------------LINEAR MODELS----------------------------------

Fit linear models for comparisons of interest.
Linear modeling in limma is carried out using the lmFit and contrasts.fit
  functions.
  
What do lm.fit and contrasts.fit do?
  1) Fit a separate model to the expression values for each gene.
  2) An empirical Bayes moderation is carried out by borrowing info across all
     the gene to obtain more precise estimates of gene-wise variability.
  3) The model's residual variances are plotted against average expression 
     values in the next figure.

It can be seen from this plot that the variance is no longer dependent on the 
  mean expression level.
  
```{r}
# Fits linear model for each gene given a series of arrays
fit <- lmFit(
  v,  # object containing log-expression values for a series of arrays
  design  # design matrix
  )

```


Contrast design for differential expression.
```{r}
contrasts <-
  makeContrasts(
    clean_vs_dirty = clean - dirty,
    levels = colnames(design)
  )

head(contrasts)
```

Assign allComparisons to a vector for later use.
```{r}
allComparisons <- colnames(contrasts)
allComparisons # check
```

Run contrast analysis.
Given a linear model fit, compute estimated coefficients and standard errors
  for a given set of contrasts.
```{r}
vfit <- contrasts.fit(fit, contrasts = contrasts)
```

Compute differential expression based on the empirical Bayes moderation of the
  standard errors towards a common value.
```{r}
veBayesFit <- eBayes(vfit)
plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")
```


Looking at N of differentially expressed genes (DEGs) with adjusted p < 0.01 
  and log2FC > 2.

decideTests() identifies which genes are significantly differentially expressed
  for each contrast from a fit object containing p-values and test statistics.
```{r}
sumTable <- 
  summary(decideTests(
    vfit,  # object
    # by default the method = "separate"
    adjust.method = "BH", 
    p.value = 0.05,
    lfc = 1  # numeric, minimum absolute log2-fold change required
  ))

head(sumTable)
```


--------------------------------------DEG TABLES--------------------------------

For each comparison...extract table of the top-ranked genes (DEGs) from a linear 
model fit AND output a table.
  
NOTE: Log2FC of 1 is equivalent to linear fold change of 2.

```{r}
coef = 1
for (i in allComparisons) {
  
  # p < 1, log2fc > 0 ----------------------------------------------------------
  vTopTableAll <-
    topTable(
      veBayesFit, 
      coef = coef,  
      n = Inf, 
      p.value = 1,
      lfc = 0 
    )
  #output as txt file
  path <- paste("DEGs_all_samples/DEG_", i, "_PVal1_Log2FC0.txt", sep = "") 
  write.table(
    vTopTableAll,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # p < 0.05, log2fc > 0 -------------------------------------------------------
  vTopTable1 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.05,
      lfc = 0
    )
  path <- paste("DEGs_all_samples/DEG_", i, "_PVal0.05_Log2FC0.txt", sep = "") 
  write.table(
    vTopTable1,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # p < 0.01, log2fc > 0 -------------------------------------------------------
  vTopTable2 <-
    topTable(  
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.01,
      lfc = 0 
    )
  path <- paste("DEGs_all_samples/DEG_", i, "_PVal0.01_Log2FC0.txt", sep = "") 
  write.table(
    vTopTable2,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # p < 0.01, log2fc > 1 -------------------------------------------------------
  vTopTable3 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.01,
      lfc = 1 
    )
  path <- paste("DEGs_all_samples/DEG_", i, "_PVal0.01_Log2FC1.txt", sep = "") 
  write.table(
    vTopTable3,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # p < 0.01, log2fc > 2 -------------------------------------------------------
  vTopTable4 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.01,
      lfc = 2 
    )
  path <- paste("DEGs_all_samples/DEG_", i, "_PVal0.01_Log2FC2.txt", sep = "") 
  write.table(
    vTopTable4,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # increment the coefficient --------------------------------------------------
  ++coef
}
```



----------------------------------VOLCANO PLOTS---------------------------------

Using the DEG tables we made, graph a volcano plot.
clean_vs_dirty = clean - dirty
```{r}
# read in table
clean_vs_dirty <-
  read.table(
    "DEGs_all_samples/DEG_clean_vs_dirty_PVal1_Log2FC0.txt",
    header = TRUE,
    sep = "\t",
    stringsAsFactors = F
  )

# create dataframe of pval, logFC, gene_name
df <- data.frame(
  clean_vs_dirty$gene_symbol,
  clean_vs_dirty$adj.P.Val,
  clean_vs_dirty$logFC
)
colnames(df) <- c("gene_name","adj.P.Val","logFC")

# extract significant genes
# subset df to have logFC >= 1, adj.P.Val <= 0.05
# Log2FC of 1 is equivalent to linear fold change of 2.
# extract gene_name column
df_sig <- df[(abs(df$logFC) >= 1 & df$adj.P.Val <= 0.05),]$gene_name

# extract non-significant genes
# subset df to exclude significant genes
df_nonsig <- subset(df, !(gene_name %in% df_sig))

# populate 4th column with the number 1 and rename the column to "Color"
df_nonsig <- cbind(df_nonsig , rep(1, nrow(df_nonsig)))
colnames(df_nonsig)[4] <- "Color"

# create list of up-regulated genes
# logFC >= 1, adj.P.val <= 0.05
upreg <- subset(df, df$logFC >= 1 & df$adj.P.Val <= 0.05  & 
                  (gene_name %in% df_sig))

# populate 4th column with the number 2 and rename the column to "Color"
upreg <- cbind(upreg, rep(2, nrow(upreg)))
colnames(upreg)[4] <- "Color"

# create list of down-regulated genes
# logFC <= 1, adj.P.val <= 0.05
downreg <-
  subset(df, df$logFC <= -1 & df$adj.P.Val <= 0.05  &
      (gene_name %in% df_sig)
  )

# populate 4th column with the number 3 and rename the column to "Color"
downreg <-
  cbind(downreg, rep(3, nrow(downreg)))
colnames(downreg)[4] <- "Color"

# create dfPlot
# join by rows (should have colors 1, 2,3 present now)
# nonSig + the up/down-regulated genes
dfPlot <- rbind(df_nonsig, upreg, downreg)

# factor the color column
dfPlot$Color <- as.factor(dfPlot$Color)

```


Now, create scatter plot of logFC vs log10(adj.P.Val).
```{r}
# Constructing the plot object.
p <-
  ggplot(data = dfPlot, aes(
    x = logFC, # x-axis is logFC
    y = -log10(adj.P.Val), # y-axis will be log10 of adj.P.Val
    color = Color)) + # color is based on factored Color column
  geom_point(alpha = 0.8, size = 3.5) + # create scatterplot, alpha makes points transparent
  theme_bw() +
  theme(legend.position = "none") + # no legend
  #xlim(c(-10, 10)) + ylim(c(0, 11)) + # x and y axis limits
  scale_color_manual(values = c("azure3", "blue")) + # set factor colors
  labs(
    title = "", # no main title
    x = expression(log[2](FC)), # x-axis title
    y = expression(-log[10] ~ "(FDR-adjusted " ~ italic("p") ~ "-value)") # y-axis title
  ) +
  theme(axis.title.x = element_text(size = 10),
        axis.text.x = element_text(size = 10)) +
  theme(axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10))
p
```


Add onto our original plot.
Add cutoff dashed lines.
```{r}
p2 <- p + geom_hline(yintercept = 1.25,
                     colour = "#000000",
                     linetype = "dashed") +
  geom_vline(xintercept = 1,
             colour = "#000000",
             linetype = "dashed") +
  geom_vline(xintercept = -1,
             colour = "#000000",
             linetype = "dashed")
p2
```



```{r}
subset_data <- subset(dfPlot, adj.P.Val < 0.05 & logFC < -1 | adj.P.Val < 0.05 & logFC > 1 )

p3 <- p2 + ggtitle("Dirty vs Clean") +
  geom_text_repel(data=subset_data,
                  aes(x = logFC, y = -log10(adj.P.Val),
                      label = gene_name), color = "black", fontface="italic")
p3
```

Ë†
```{r}
subset_data
```

```{r}
df[order(df$adj.P.Val, decreasing = TRUE),]
```


